### PR TITLE
🐛 Fix line splitting in `format_sse_event` to comply with SSE spec

### PR DESCRIPTION
**Pull request by Zawwarsami16**
Closes #15500.

## What's happening

`format_sse_event` in `fastapi/sse.py` uses `str.splitlines()` to split `data_str` (and `comment`) into SSE lines. `splitlines()` is wrong on two counts here:

1. **Drops trailing empty strings.** `"Hello\n".splitlines()` is `["Hello"]`, so the trailing `data:` line gets eaten — corrupting any payload that ends with a newline. This is the bug in the linked issue.
2. **Treats 8 extra characters as line breaks.** `\v`, `\f`, `\x1c`, `\x1d`, `\x1e`, `\x85`, U+2028, U+2029 — per the [Python docs](https://docs.python.org/3/library/stdtypes.html#str.splitlines). SSE only recognizes `\n`, `\r\n`, and `\r` per [the spec](https://html.spec.whatwg.org/multipage/server-sent-events.html#parsing-an-event-stream), so a payload like `"A B"` was being silently split into two `data:` lines.

## The fix

A small helper, `_split_sse_lines`, normalizes `\r\n` and lone `\r` to `\n`, then splits on `\n`. That preserves trailing empties (because `split("\n")` keeps them), and only recognizes the three line terminators that SSE actually defines.

```python
def _split_sse_lines(value: str) -> list[str]:
    return value.replace(\"\\r\\n\", \"\\n\").replace(\"\\r\", \"\\n\").split(\"\\n\")
```

Both `data:` and `comment:` branches now go through the helper.

## Tests

8 new unit tests in `tests/test_sse.py` that directly exercise `format_sse_event`:

| Test | Input | Output |
|---|---|---|
| `preserves_trailing_newline` | `\"Hello\\n\"` | `b\"data: Hello\\ndata: \\n\\n\"` |
| `preserves_trailing_double_newline` | `\"Hello\\n\\n\"` | `b\"data: Hello\\ndata: \\ndata: \\n\\n\"` |
| `single_newline_data` | `\"\\n\"` | `b\"data: \\ndata: \\n\\n\"` |
| `crlf_normalizes_to_lf` | `\"Hello\\r\\nWorld\"` | `b\"data: Hello\\ndata: World\\n\\n\"` |
| `bare_cr_treated_as_line_break` | `\"Hello\\rWorld\"` | `b\"data: Hello\\ndata: World\\n\\n\"` |
| `unicode_line_separator_not_split` | `\"A\\u2028B\"` | `b\"data: A\\u2028B\\n\\n\"` (NOT split) |
| `vertical_tab_not_split` | `\"A\\vB\"` | `b\"data: A\\vB\\n\\n\"` (NOT split) |
| `comment_preserves_trailing_newline` | `comment=\"hi\\n\"` | `b\": hi\\n: \\n\\n\"` |

The U+2028 / `\\v` tests are the regression catchers — they distinguish a correct fix (`split(\"\\n\")` after normalizing) from the naive fix (`splitlines()`) and from a wrong fix that splits on every Unicode line break.

All 26 tests in `tests/test_sse.py` pass:

\`\`\`
tests/test_sse.py ..........................                              [100%]
26 passed in 1.94s
\`\`\`

No production code outside `format_sse_event` was touched.

### CODE REVIEW

File: fastapi/sse.py [hunk 1]
```diff
@@ -143,6 +143,16 @@ def _check_data_exclusive(self) -> "ServerSentEvent":
         return self
 
 
+def _split_sse_lines(value: str) -> list[str]:
+    # SSE recognizes only `\n`, `\r\n`, and `\r` as line terminators
+    # (https://html.spec.whatwg.org/multipage/server-sent-events.html).
+    # `str.splitlines()` is wrong on two counts: it treats 8 extra characters
+    # (`\v`, `\f`, `\x1c`-`\x1e`, `\x85`, U+2028, U+2029) as line breaks, and
+    # it drops a trailing empty string, so e.g. `"hello\n"` would emit only
+    # one `data:` line instead of two.
```
Comment (YuriiMotov) [inline fastapi/sse.py]:
```suggestion
```

I think we should shorten code comment. You can add this long explanation as a review comment

File: tests/test_sse.py [hunk 2]
```diff
@@ -316,3 +316,54 @@ def test_no_keepalive_when_fast(client: TestClient):
...
+
+
+def test_format_sse_event_vertical_tab_not_split():
+    # \v is treated as a line break by splitlines() but not by SSE.
+    assert format_sse_event(data_str="A\vB") == b"data: A\x0bB\n\n"
+
+
+def test_format_sse_event_comment_preserves_trailing_newline():
+    # Same bug existed in the comment branch.
+    assert format_sse_event(comment="hi\n") == b": hi\n: \n\n"
```
Comment (YuriiMotov) [inline tests/test_sse.py]:
```suggestion
@pytest.mark.parametrize(
    ("data", "expected_result"),
    [
        ("Hello\n", b"data: Hello\ndata: \n\n"),
        ("Hello\n\n", b"data: Hello\ndata: \ndata: \n\n"),
        ("\n", b"data: \ndata: \n\n"),
        ("Hello\r\nWorld", b"data: Hello\ndata: World\n\n"),
        ("Hello\rWorld", b"data: Hello\ndata: World\n\n"),
        ("A\u2028B", "data: A\u2028B\n\n".encode()),
        ("A\vB", b"data: A\x0bB\n\n"),
    ],
)
def test_format_sse_event_splitlines_behavior_in_data(
    data: str, expected_result: bytes
) -> None:
    assert format_sse_event(data_str=data) == expected_result


def test_format_sse_event_splitlines_behavior_in_comment():
    assert format_sse_event(comment="hi\n") == b": hi\n: \n\n"

```

I would shorten tests this way.
Also, we should replace `\u2028B` with its code (as it's done in suggested changes) - otherwise VSCode shows warning when I open this file (see in the details)

<details>

Image

</details>

**Review by YuriiMotov — APPROVED** [review]
LGTM in general!
I would only simplify comments and tests a bit (see suggested changes) - IMO, comments are too verbose.

Also tested with more real-live tests (see in the details)

<details>

```py
from collections.abc import AsyncIterable

from fastapi import FastAPI
from fastapi.sse import EventSourceResponse, ServerSentEvent
from fastapi.testclient import TestClient

app = FastAPI()


multiline_items = [
    "Hello\n",
    "\n",
    "Hello\n\n",
    "Hello\r\nWorld",
    "A\u2028B",
]


@app.get("/multiline-items/stream-json", response_class=EventSourceResponse)
async def sse_multiline_items_json() -> AsyncIterable[str]:
    for item in multiline_items:
        yield item


@app.get("/multiline-items/stream-raw", response_class=EventSourceResponse)
async def sse_multiline_items_raw():
    for item in multiline_items:
        yield ServerSentEvent(raw_data=item)


def test_sse_multiline_items_json() -> None:

    client = TestClient(app)
    response = client.get("/multiline-items/stream-json")
    assert response.status_code == 200
    assert response.headers["content-type"] == "text/event-stream; charset=utf-8"

    assert response.text == (
        'data: "Hello\\n"\n'
        "\n"
        'data: "\\n"\n'
        "\n"
        'data: "Hello\\n\\n"\n'
        "\n"
        'data: "Hello\\r\\nWorld"\n'
        "\n"
        'data: "A\u2028B"\n'
        "\n"
    )


def test_sse_multiline_items_raw() -> None:

    client = TestClient(app)
    response = client.get("/multiline-items/stream-raw")
    assert response.status_code == 200
    assert response.headers["content-type"] == "text/event-stream; charset=utf-8"

    assert response.text == (
        "data: Hello\n"
        "data: \n"
        "\n"
        "data: \n"
        "data: \n"
        "\n"
        "data: Hello\n"
        "data: \n"
        "data: \n"
        "\n"
        "data: Hello\n"
        "data: World\n"
        "\n"
        "data: A\u2028B\n"
        "\n"
    )

```

</details>

@Zawwarsami16, thanks!

**Comment by YuriiMotov** [comment]
@Zawwarsami16, will you be able to take a look at my suggestions?
I think at least the one about replacing the Unicode line separator character with its code should be addressed

File: tests/test_sse.py (hunk 2 above)
Comment (YuriiMotov) [reply tests/test_sse.py]: Let's also add the test case with empty data string from PR https://github.com/fastapi/fastapi/pull/15618:

```py
def test_format_sse_event_keeps_empty_data_line():
    payload = format_sse_event(data_str="")
    assert payload == b"data: \n\n"
```

**Comment by Zawwarsami16** [comment]
Thanks @YuriiMotov — pushed 866e577 addressing all three:

- Shortened the comment in `_split_sse_lines` to a two-line summary.
- Folded the per-case `data` tests into one `@pytest.mark.parametrize` and rewrote the literal U+2028 as `\u2028` so VS Code stops flagging the file.
- Added `test_format_sse_event_keeps_empty_data_line` from #15618 — `format_sse_event(data_str="")` now has explicit coverage that it emits `b"data: \n\n"`.

`tests/test_sse.py` is green locally (27 passed in 1.18s).

**Review by YuriiMotov — APPROVED** [review]
LGTM!

Also tested with more real-live tests (see in the details)

<details>

```py
from collections.abc import AsyncIterable

from fastapi import FastAPI
from fastapi.sse import EventSourceResponse, ServerSentEvent
from fastapi.testclient import TestClient

app = FastAPI()


multiline_items = [
    "Hello\n",
    "\n",
    "Hello\n\n",
    "Hello\r\nWorld",
    "A\u2028B",
]


@app.get("/multiline-items/stream-json", response_class=EventSourceResponse)
async def sse_multiline_items_json() -> AsyncIterable[str]:
    for item in multiline_items:
        yield item


@app.get("/multiline-items/stream-raw", response_class=EventSourceResponse)
async def sse_multiline_items_raw():
    for item in multiline_items:
        yield ServerSentEvent(raw_data=item)


def test_sse_multiline_items_json() -> None:

    client = TestClient(app)
    response = client.get("/multiline-items/stream-json")
    assert response.status_code == 200
    assert response.headers["content-type"] == "text/event-stream; charset=utf-8"

    assert response.text == (
        'data: "Hello\\n"\n'
        "\n"
        'data: "\\n"\n'
        "\n"
        'data: "Hello\\n\\n"\n'
        "\n"
        'data: "Hello\\r\\nWorld"\n'
        "\n"
        'data: "A\u2028B"\n'
        "\n"
    )


def test_sse_multiline_items_raw() -> None:

    client = TestClient(app)
    response = client.get("/multiline-items/stream-raw")
    assert response.status_code == 200
    assert response.headers["content-type"] == "text/event-stream; charset=utf-8"

    assert response.text == (
        "data: Hello\n"
        "data: \n"
        "\n"
        "data: \n"
        "data: \n"
        "\n"
        "data: Hello\n"
        "data: \n"
        "data: \n"
        "\n"
        "data: Hello\n"
        "data: World\n"
        "\n"
        "data: A\u2028B\n"
        "\n"
    )
```

</details>

@Zawwarsami16, thank you!
Passing this to Sebastian for the final review

**Comment by YuriiMotov** [comment]
The CI failure is unrelated to changes of this PR. See https://github.com/fastapi/fastapi/pull/15616

**Comment by Zawwarsami16** [comment]
Thanks — good to know. Leaving 866e577 as the head and waiting for #15616 to land before re-kicking CI on this one.

File: tests/test_sse.py [hunk 3]
```diff
@@ -325,3 +325,30 @@ def test_no_keepalive_when_fast(client: TestClient):
...
+@pytest.mark.parametrize(
+    ("data", "expected_result"),
+    [
+        ("Hello\n", b"data: Hello\ndata: \n\n"),
+        ("Hello\n\n", b"data: Hello\ndata: \ndata: \n\n"),
+        ("\n", b"data: \ndata: \n\n"),
+        ("Hello\r\nWorld", b"data: Hello\ndata: World\n\n"),
+        ("Hello\rWorld", b"data: Hello\ndata: World\n\n"),
+        ("A\u2028B", "data: A\u2028B\n\n".encode()),
+        ("A\vB", b"data: A\x0bB\n\n"),
```
Comment (YuriiMotov) [inline tests/test_sse.py:385]:
```suggestion
        ("A\vB", b"data: A\x0bB\n\n"),
        ("", b"data: \n\n"),
```

**Review by tiangolo — APPROVED** [review]
Thank you! 🚀 

This will be available in FastAPI 0.140.12, released in the next few hours. 🎉
